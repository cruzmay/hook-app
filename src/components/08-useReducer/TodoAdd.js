import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ({ handleAddToDo }) => {
    
    const [{description}, handleInputChange, reset] = useForm({description: ''})

    const handleSubmit = (e) => {
        e.preventDefault()

        if(description.trim().length <= 1){
            return
        }

        const newTodo = {
            id : new Date().getTime(),
            desc: description,
            done: false
        }

        handleAddToDo(newTodo)
        reset()

    }

    return (
        <>
            <h3>Agregar</h3>
            <form 
                onSubmit={handleSubmit}
                className="form-group col-12 m-4 d-flex justify-content-center"
            >
                <input
                    className="col-9"
                    type="test"
                    name="description"
                    placeholder="aprender"
                    autoComplete="off"
                    value={description}
                    onChange={handleInputChange}
                />
                <button
                    type="submit"
                    className="btn btn-primary col-3 ms-4"
                >
                    Agregar
                </button>
            </form>
        </>
    )
}
