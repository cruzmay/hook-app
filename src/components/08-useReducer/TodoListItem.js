import React from 'react'

export const TodoListItem = ({id, desc, done, handleDelete, handleToogle, index}) => {
    return (
        <li 
        
        className="list-group-item col-12 d-flex justify-content-between align-items-center"
        >
            <p 
                className={`col-4 m-0 task ${ done && 'tacha'} `}
                onClick={()=>handleToogle(id)}
            >
            {index + 1}.- {desc}
            </p>
            <button 
                className="btn btn-small btn-danger"
                onClick={() => handleDelete(id)}
            >
                Borrar
            </button>
        </li>
    )
}
