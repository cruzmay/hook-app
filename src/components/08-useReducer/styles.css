body {
  padding: 50px;
}
.task {
  cursor: pointer;
}
.tacha {
  text-decoration: line-through;
}
