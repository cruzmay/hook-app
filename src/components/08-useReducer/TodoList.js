import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ({todos, handleDelete, handleToogle}) => {
    return (
        <ul>
            {
                todos.map((todo, i)=> 
                    <TodoListItem 
                        {...todo}
                        key={todo.id}
                        handleDelete={handleDelete}
                        handleToogle={handleToogle}
                        index = {i}
                    />
                )
            }
            
        </ul>
    )
}
