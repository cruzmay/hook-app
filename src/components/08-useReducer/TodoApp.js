import React, { useEffect, useReducer } from 'react';
import './styles.css'
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

 const init = () => {
     return JSON.parse(localStorage.getItem('todos'))
 }

export const TodoApp = () => {
    
    const [todos, dispatch] = useReducer(todoReducer, [], init)

    useEffect(()=>{
        localStorage.setItem('todos', JSON.stringify(todos))
    },[todos])

    const handleDelete = (todoId) => {
        console.log(todoId)
        const deleteAction = {
            type: 'delete',
            payload: todoId
        }
        dispatch(deleteAction)
    }

    const handleToogle = (todoId) => {
        // console.log(todoId)
        dispatch({
            type: 'toogle',
            payload: todoId
        })
    }

    const handleAddToDo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        })
    }

    return (
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr/>
            <TodoAdd
                handleAddToDo={handleAddToDo}
            />
            <TodoList
                todos = {todos}
                handleDelete={handleDelete}
                handleToogle={handleToogle}
            />
        </div>
    )
}
