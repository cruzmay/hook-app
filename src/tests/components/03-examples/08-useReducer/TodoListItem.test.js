import React from 'react';
import { shallow } from "enzyme"
import { TodoListItem } from "../../../../components/08-useReducer/TodoListItem"
import { demoTodos } from '../../../fixtures/demoTodos';

describe('pruebas en <TodoListItem/>', () => {

    const handleDelete = jest.fn()
    const handleToogle = jest.fn()

    const wrapper = shallow(<TodoListItem
            id={demoTodos[1].id}
            desc={demoTodos[1].desc}
            done={false}
            key={demoTodos[1].id}
            handleDelete={handleDelete}
            handleToogle={handleToogle}
            index={1}
        />)

    test('debe de mostrarse correctamente', () => {
      //snapshot  
        expect(wrapper).toMatchSnapshot()
    })
    test('debe de llamar la funcion handleDelete', () => {
        //jest.fn()
        wrapper.find('button').simulate('click')
        //toHaveBeenCalled
        expect(handleDelete).toHaveBeenCalled()
        //toHaveBeenCalledWith
        expect(handleDelete).toHaveBeenCalledWith(demoTodos[1].id)
    })
    test('debe de llamar la funcion handleToggle', () => {
        //jest.fn()
        wrapper.find('.task').simulate('click')
        //toHaveBeenCalled
        expect(handleToogle).toHaveBeenCalled()
        //toHaveBeenCalledWith
        expect(handleToogle).toHaveBeenCalledWith(demoTodos[1].id)
    })
    test('debe de mostrar el texto correctamente', () => {
        //contenido del parrafo
        console.log(wrapper.find('p').text())
        expect(wrapper.find('p').text()).toBe(`${1 + 1}.- ${demoTodos[1].desc}`)
    })
    test('debe de marcarse el texto en handleToogle', () => {
        
        const wrapper = shallow(<TodoListItem done={true}/>)
        expect(wrapper.find('p').hasClass('tacha')).toBe(true)

    })
    
})
