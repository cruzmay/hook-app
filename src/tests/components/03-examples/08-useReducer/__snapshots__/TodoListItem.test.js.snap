// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <TodoListItem/> debe de mostrarse correctamente 1`] = `
<li
  className="list-group-item col-12 d-flex justify-content-between align-items-center"
>
  <p
    className="col-4 m-0 task false "
    onClick={[Function]}
  >
    2
    .- 
    Aprender Mongo
  </p>
  <button
    className="btn btn-small btn-danger"
    onClick={[Function]}
  >
    Borrar
  </button>
</li>
`;
